File: conventional-commits-common/src/main/java/com/smartling/cc4j/semantic/release/common/SimpleSemanticVersionChangeResolver.java
package com.smartling.cc4j.semantic.release.common;

import org.eclipse.jgit.revwalk.RevCommit;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

public class SimpleSemanticVersionChangeResolver implements SemanticVersionChangeResolver
{
    @Override
    public SemanticVersionChange resolveChange(final Iterable<RevCommit> commits)
    {
        Objects.requireNonNull(commits, "commits may not be null");
        final List<Commit> commitList = new ArrayList<>();
        commits.iterator().forEachRemaining(c -> commitList.add(new Commit(new GitCommitAdapter(c))));
        SemanticVersionChange change = SemanticVersionChange.NONE;

        for (final Commit c : commitList)
        {
            final Optional<ConventionalCommitType> commitType = c.getCommitType();
            if (commitType.isPresent())
            {
                if (SemanticVersionChange.MAJOR.equals(commitType.get().getChangeType()))
                {
                    return SemanticVersionChange.MAJOR;
                }
                else if (SemanticVersionChange.MINOR.equals(commitType.get().getChangeType()))
                {
                    change = SemanticVersionChange.MINOR;
                }
                else if (SemanticVersionChange.PATCH.equals(commitType.get().getChangeType()))
                {
                    if (!SemanticVersionChange.MINOR.equals(change))
                    {
                        change = SemanticVersionChange.PATCH;
                    }
                }
            }
        }

        return change;
    }
}

File: conventional-commits-common/src/main/java/com/smartling/cc4j/semantic/release/common/SimpleSemanticVersionChangeResolver.kt
package com.smartling.cc4j.semantic.release.common

import org.eclipse.jgit.revwalk.RevCommit

class SimpleSemanticVersionChangeResolver : SemanticVersionChangeResolver {
    override fun resolveChange(commits: Iterable<RevCommit>): SemanticVersionChange {
        return commits
            .map { c -> Commit(GitCommitAdapter(c)) }
            .fold(SemanticVersionChange.NONE) { type, commit ->
                val changeType = this.resolveChange(commit)
                return when {
                    SemanticVersionChange.MAJOR == changeType -> SemanticVersionChange.MAJOR
                    SemanticVersionChange.MINOR == changeType && SemanticVersionChange.MAJOR != type -> SemanticVersionChange.MINOR
                    SemanticVersionChange.PATCH == changeType && SemanticVersionChange.MINOR != type -> SemanticVersionChange.PATCH
                    else -> type
                }
            }
    }

    fun resolveChange(commit: Commit): SemanticVersionChange {
        val changeType = commit.getCommitType()?.changeType
        return when {
            SemanticVersionChange.MAJOR == changeType -> SemanticVersionChange.MAJOR
            SemanticVersionChange.MINOR == changeType -> SemanticVersionChange.MINOR
            SemanticVersionChange.PATCH == changeType -> SemanticVersionChange.PATCH
            else -> SemanticVersionChange.NONE
        }
    }
}
